### PR TITLE
Implementation of nosemgrep tag in OCaml

### PR DESCRIPTION
**Pull request by dinosaure**
PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [ ] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [ ] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [ ] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

This PR wants to implement the `nosemgrep` tag into the `osemgrep` OCaml code as the Python implemention (`semgrep`). Currently, from my handmade tests, the `nosemgrep` tag works. However, the end-to-end test fails due to some other things unrelated to this PR (like absolute path of files into the `scanned` JSON field).

Some questions persists however:
- [ ] do we have a function to _trim_ ID collected?
- [ ] my usage of `cli_error` is correct?
- [ ] `nosem_line_comment_re` seems unused by the current Python implemention, should I delete it?

I make this PR as a draft because I'm still unsure that it's ready to merge.

### CODE REVIEW

File: src/osemgrep/reporting/Nosem.ml [hunk 1]
```diff
@@ -35,14 +35,15 @@ let rule_id_re_str = {|(?:[:=][\s]?(?P<ids>([^,\s](?:[,\s]+)?)+))?|}
...
-let _nosem_inline_re = SPcre.regexp nosem_inline_re_str ~flags:[ `CASELESS ]
+let nosem_inline_re = SPcre.regexp nosem_inline_re_str ~flags:[ `CASELESS ]
 
 (*
    As a hack adapted from semgrep-agent,
    we assume comment markers are one of these special characters
 *)
 let _nosem_inline_comment_re =
   SPcre.regexp (spf {|[:#/]+%s$|} nosem_inline_re_str) ~flags:[ `CASELESS ]
+(* XXX(dinosaure): seems unused by the last version of `cli/` *)
```
Comment (aryx) [inline src/osemgrep/reporting/Nosem.ml]: yes, feel free to remove the dead code here and also in cli/.

File: src/osemgrep/reporting/Nosem.ml [hunk 2]
```diff
@@ -69,11 +70,96 @@ let _nosem_previous_line_re =
...
 
-(* TODO: probably need to pass some flags from the CLI *)
-let process_ignores (out : Out.cli_output) : Out.cli_output =
-  (* TODO: port nosemgrep.py, filter in out.matches the
-     matches where the match has a corresponding nosem: annotation in
-     the corresponding file, and add new errors if nosem: has the wrong
-     format in the file.
-  *)
-  out
+let rule_match_nosem ~strict (rule_match : Out.cli_match) :
```
Comment (aryx) [inline src/osemgrep/reporting/Nosem.ml:78]: you can move this in the Helpers section (it's not an entry point of the module).

File: src/osemgrep/reporting/Nosem.ml [hunk 3]
```diff
@@ -69,11 +70,96 @@ let _nosem_previous_line_re =
...
+
+  let previous_line, line =
+    match lines with
+    | line0 :: line1 :: _ when rule_match.Out.start.line > 0 ->
+        (Some line0, Some line1)
+    | line :: _ -> (None, Some line)
+    | [] (* XXX(dinosaure): is it possible? *) -> (None, None)
+  in
+
+  let recognise_and_collect ~rex line =
```
Comment (aryx) [inline src/osemgrep/reporting/Nosem.ml]: avoid nested functions. Move out in Helper section, and give type and small explanation if needed.

File: src/osemgrep/reporting/Nosem.ml [hunk 4]
```diff
@@ -69,11 +70,96 @@ let _nosem_previous_line_re =
...
+    Result.map
+      (Array.map (fun subst ->
```
Comment (aryx) [inline src/osemgrep/reporting/Nosem.ml]: We tend to favor using |> when we can, to be able to 
see first the data structure we operate on, and then the actuall code, so here
like (SPcre.exec_all ~rex line |> Array.map (fun subst ...)

But if you feel strongly about not using |>, that's fine, some of us don't.

File: src/osemgrep/reporting/Nosem.ml [hunk 5]
```diff
@@ -69,11 +70,96 @@ let _nosem_previous_line_re =
...
+           try Some (Pcre.get_named_substring rex "ids" subst) with
+           | _ -> None))
+      (SPcre.exec_all ~rex line)
+    |> Result.to_option
+  in
+
+  let no_ids = Array.for_all Option.is_none in
+
+  let ids_line, ids_previous_line =
+    ( Option.bind line (recognise_and_collect ~rex:nosem_inline_re),
```
Comment (aryx) [inline src/osemgrep/reporting/Nosem.ml]: I never used Option.bind, but as a general rule with option we sometimes prefer to be explicit
and use match x | Some -> .. None -> ...
especially if the resulting code is actually not that longer than using Option.map, Option.bind.

**Review by aryx — APPROVED** [review]

**Review by aryx — CHANGES_REQUESTED** [review]
Some small requests, but overall looks good.

**Review by aryx — APPROVED** [review]

File: libs/commons/File.mli [hunk 6]
```diff
@@ -126,3 +126,12 @@ val erase_this_temp_file : Fpath.t -> unit
...
 val filesize : Fpath.t -> int
+
+(* [lines_of_file (start_line, end_line) file] returns
+ * the list of lines from start_line to end_line included.
+ *
+ * Note that the returned lines do not contain \n.
+ *
+ * This function is slow, you should not use it!
+ *)
+val lines_of_file : int * int -> Common.filename -> string list
```
Comment (mjambon) [inline libs/commons/File.mli:137]: This module (which was created recently) should use only `Fpath.t` for file paths. We're moving away from `Common.filename` which is an alias for `string`. We haven't finished migrating all our code to `Fpath`, so if it's too inconvenient to use the fpath version of this function, move the non-fpath version to the `Common` module and have a wrapper using fpath in the `File` module.

**Pull request merged**